pandas/core/ops.py

```diff
@@ -990,6 +991,26 @@ def _construct_divmod_result(left, result, index, name, dtype):
     )
 
 
+def dispatch_to_extension_op(op, left, right):
+    """
+    Assume that left or right is a Series backed by an ExtensionArray,
+    apply the operator defined by op.
+    """
+
+    # The op calls will raise TypeError if the op is not defined
+    # on the ExtensionArray
+    if is_extension_array_dtype(left):
+        res_values = op(left.values, right)
+    else:
+        # We know that left is not ExtensionArray and is Series and right is
+        # ExtensionArray.  Want to force ExtensionArray op to get called
+        res_values = op(list(left.values), right.values)
```

Comment:
was this needed to fix failing tests? (this was not here in a previous version I think? and eg the dispatch to index is only dealing with the left case)

Reply:
@jbrockmendel Wrote:
> >  Another possible implementation might be to add a test in the wrappers that says that if is_extension_dtype(right) return NotImplemented

>This is much closer to what we've had in mind with the recent refactoring in ops. Would this look something like splitting up the conditions on 1082-1085? We might need to up-cast right to be a Series in order for the deference chain to go through; I'll need to double-check.

Yes, I think it would be something like:
```
       elif is_extension_array_dtype(left):
           return dispatch_to_extension_op(op, left, right)
       elif (is_extension_array_dtype(right) and                 
              not is_categorical_dtype(right)):
           return NotImplemented
```
But I would have to test the concept.  I don't think the upcast is necessary. We know that left is a Series. So python would then see that `op(left, right)` returns `NotImplemented`, and then try to call `reverse_op(right, left)`, so that the reverse operator is called, but using the implementation of `right`.  

I'm not going to try making that change without feedback from others.

> Why is the list(left.values) necessary? I would have thought left.values would get the job done.

Because `left.values` could be any type (numpy.ndarray, pd.Categorical, etc.), and we don't want the version of op to be called that is associated with the class of left.values . So by making it a list, we are forcing the reverse operator to be called on right.values, which is the EA implementation.

Reply:
> Another possible implementation might be to add a test in the wrappers that says that if is_extension_dtype(right) return NotImplemented, which will then make python call the reverse operator.

I think this is the better way to go. 

For me it is fine to not yet do this in this PR, but then I would also not include the above change (and live with that the first iteration in this PR is not yet ideal for cases where the ExtensionArray is the right variable)

Reply:
@jorisvandenbossche If I don't handle the  case where ExtensionArray is the right variable now, then my tests will fail.  So I either implement it the other way (as you quoted above), or leave it as I have it now.  Should I make that change (using `NotImplemented`)?

Reply:
> Should I make that change (using NotImplemented)?

I would at least try this

Reply:
@jorisvandenbossche Yes, I will try this in a new PR.

Reply:
@jorisvandenbossche (and @jbrockmendel) I tried the `NotImplemented` idea and it doesn't work.  I think the reason is as follows.  Let's say that we have one `Series` called `s1` that has a normal (`int` or `double`) dtype. And `s2` has a dtype of the EA.  Let's say you write the expression `s1 + s2`.  Python sees that you are adding two `Series` objects.  So internally, if we return `NotImplemented` because `s2` is an EA, then Python will say "Well, you don't know how to add a Series to a Series, so I'm not going to bother calling the reverse operator."  In other words, the reverse operator is only called when the two types passed to the operator are different, and `NotImplemented` is returned when applying the operator to the first argument.

Reply:
I'm not entirely sure I follow.  Are you saying that both `s1.__add__(s2)` and `s2.__add__(s1)` return `NotImplemented`?

The dispatch logic should do something like
```
out = s1.values + s2
return Series(out, name=...)
```

It is `s1.values.__add__` that should return `NotImplemented` in this case.  Is this not an option for some reason?

Reply:
@jbrockmendel I've tried two different alternatives (regarding replacing this code
```python
else:
        # We know that left is not ExtensionArray and is Series and right is
        # ExtensionArray.  Want to force ExtensionArray op to get called
        res_values = op(list(left.values), right.values)
```
that this discussion refers to). Note the section of code is hit when `left.values` is *not* an EA, but `right.values` is an EA.
1. Use
```python
else:
    return NotImplemented
```
2. Use
```python
else:
    res_values = op(left.values, right.values)
```

In the first case, python sees that `s1.__add__(s2)` is `NotImplemented`, and because `s1` and `s2` are both of type `Series`, then `s2.__radd__(s1)` is never called.

In the second case, `left.values` is a `numpy` array, so then the `numpy` `__add__` method is called, and it is perfectly happy to return an array of `dtype` that is set to `object` rather than the dtype of the EA.  `numpy` ends up calling the `__add__` method for each of the elements.